src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/CollectionBackedStoreProvider.cs

```diff
@@ -4,167 +4,74 @@
 
 using System;
 using System.Collections.Generic;
-using System.Collections.ObjectModel;
 using System.Diagnostics;
-using System.Security.Cryptography;
 using System.Security.Cryptography.X509Certificates;
-using Microsoft.Win32.SafeHandles;
+using System.Threading;
 
 namespace Internal.Cryptography.Pal
 {
     internal class CollectionBackedStoreProvider : IStorePal
     {
-        private readonly X509Certificate2[] _certs;
-        private static readonly SafeEvpPKeyHandle InvalidPKeyHandle = new SafeEvpPKeyHandle(IntPtr.Zero, false);
+        private readonly bool _cloneOnRead;
+        private X509Certificate2[] _certs;
 
-        internal CollectionBackedStoreProvider(X509Certificate2 cert)
+        internal CollectionBackedStoreProvider(X509Certificate2 cert, bool cloneOnRead)
         {
+            _cloneOnRead = cloneOnRead;
             _certs = new X509Certificate2[] { cert };
         }
 
-        internal CollectionBackedStoreProvider(X509Certificate2Collection certs)
+        internal CollectionBackedStoreProvider(X509Certificate2Collection certs, bool cloneOnRead)
         {
+            _cloneOnRead = cloneOnRead;
             _certs = new X509Certificate2[certs.Count];
             certs.CopyTo(_certs, 0);
         }
 
-        public void Dispose()
+        internal CollectionBackedStoreProvider(List<X509Certificate2> certs, bool cloneOnRead)
         {
+            _cloneOnRead = cloneOnRead;
+            _certs = certs.ToArray();
         }
 
-        public byte[] Export(X509ContentType contentType, string password)
+        public void Dispose()
         {
-            switch (contentType)
-            {
-                case X509ContentType.Cert:
-                    return ExportX509Der();
-                case X509ContentType.Pfx:
-                    return ExportPfx(password);
-                case X509ContentType.Pkcs7:
-                    return ExportPkcs7();
-                case X509ContentType.SerializedCert:
-                case X509ContentType.SerializedStore:
-                    throw new PlatformNotSupportedException(SR.Cryptography_Unix_X509_SerializedExport);
-                default:
-                    throw new CryptographicException(SR.Cryptography_X509_InvalidContentType);
-            }
+            // Dispose is explicitly doing nothing here.
+            // The LM\Root and LM\CA stores are reused after being Disposed, because there's no
+            // point in re-allocating the array to instantiate them every time.
+            // But the interface requires Dispose.
```

Comment:
My initial inclination when I was looking at this previously was to ask whether all of the _certs could be disposed of here.  I believe the answer was "no, they may have been handed out and be in use elsewhere... might be worth noting that here.